src/app/treelist-checkbox/treelist-checkbox.component.ts

```diff
@@ -0,0 +1,250 @@
+      }
+    } else if (datasetNode?.children.length > 0) {
+      this.getAllChildren(datasetNode.children, new Set(), true).forEach(child => {
+        this.changeCheckStatusEvent.emit($event.target.id);
+        this.allStudies.delete(child);
+      });
+      this.allStudies.delete($event.target.id);
+    } else if (this.allStudies.has($event.target.id)) {
+      this.changeCheckStatusEvent.emit($event.target.id);
+      this.allStudies.delete($event.target.id);
+    }
+    if (this.allStudies.has($event.target.id)) {
+      this.changeCheckStatusEvent.emit($event.target.id);
+      this.allStudies.add($event.target.id);
+    } else {
+      this.changeCheckStatusEvent.emit($event.target.id);
+      this.allStudies.delete($event.target.id);
+    }
+    this.updateFilterStates(this.data);
+  }
+
+  public updateParents(datasetNode): void {
+    const parents = this.getAllParents(datasetNode);
+    parents.forEach((parent: string) => {
+      if (parent === this.data.dataset.name) {
+        return;
+      }
+      let flag = 0;
+      this.findNodeById(this.data, parent)?.children?.forEach(child => {
+        if (this.allStudies.has(child.dataset.name)) {
+          flag += 1;
+        }
+      });
+      if (flag === parents.length) {
+        flag = 2;
+      } else if (flag !== 0 && flag < parents.length) {
+        flag = 1;
+      }
+      switch (flag) {
+        case 0: {
+          const elements = document.querySelectorAll(`[id='${parent}']`);
+          elements.forEach(element => {
+            (element as HTMLInputElement).checked = false;
+            (element as HTMLInputElement).indeterminate = false;
+          });
+          break;
+        }
+        case 1: {
+          const elements = document.querySelectorAll(`[id='${parent}']`);
+          elements.forEach(element => {
+            (element as HTMLInputElement).checked = true;
+            (element as HTMLInputElement).indeterminate = true;
+          });
+          break;
+        }
+        case 2: {
+          const elements = document.querySelectorAll(`[id='${parent}']`);
+          elements.forEach(element => {
+            (element as HTMLInputElement).checked = true;
+            (element as HTMLInputElement).indeterminate = false;
+          });
+          break;
+        }
+      }
+    });
+  }
+
+  public updateFilterStates(data: DatasetNode): void {
+    let checkState = 0;
+    if (data.children.length > 0) {
+      data.children.forEach(child => {
+        if (child.children.length > 0) {
+          this.updateFilterStates(child);
+        }
+      });
+    }
+
+    const children = this.getAllChildren(data.children, new Set());
+    this.getAllChildren(data.children, new Set()).forEach(child => {
+      if (this.allStudies.has(child)) {
+        checkState = 1;
+      }
+    });
+    let flag = 1;
+    children.forEach(child => {
+      if (!this.allStudies.has(child)) {
+        flag = 0;
+      }
+    });
+    if (flag) {
+      checkState = 2;
+    }
```

Comment:
Minor nitpick - you could utilise [Array.some](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some) here

Reply:
Cannot be used here because it is Set, otherwise would be useful, particularly in this case